### DISCUSSION TITLE
Problem with passing query params to route.

### DISCUSSION THREAD

**Discussion by chavdar94** [fastapi/fastapi#11830] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#main.py
"""
Boot FastApi app
"""

from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.middleware.sessions import SessionMiddleware

from middleware import log_middleware
from src.api.api import api_router
from src.core.config import get_settings

settings = get_settings()


def _setup_cors(p_app: FastAPI) -> None:
    """
    Set all CORS enabled origins
    """
    p_app.add_middleware(
        CORSMiddleware,
        allow_origins=settings.BACKEND_CORS_ORIGINS,
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )


def _create_app() -> FastAPI:
    app_ = FastAPI(
        title=settings.PROJECT_NAME,
        openapi_url=f"{settings.API_V1_STR}/openapi.json",
        docs_url="/swagger",
    )
    app_.include_router(
        api_router,
        prefix=get_settings().API_V1_STR,
    )

    app_.add_middleware(SessionMiddleware, secret_key=settings.SECRET_KEY)
    app_.add_middleware(BaseHTTPMiddleware, dispatch=log_middleware)

    return app_


app = _create_app()
_setup_cors(app)
```
```#/src/api/api.py
from fastapi import APIRouter

from .routes import auth, products, url_clicks, categories

api_router = APIRouter()

api_router.include_router(auth.router, prefix="/auth", tags=["auth"])
api_router.include_router(products.router, prefix="/products", tags=["products"])
api_router.include_router(url_clicks.router, prefix="/url_clicks", tags=["url_clicks"])
api_router.include_router(categories.router, prefix="/categories", tags=["categories"])
```

```py
#/src/api/routes/products.py

from fastapi import APIRouter, Depends
from src.schemas.products import Pagination, ProductBase, ProductResponse
from sqlalchemy.orm import Session
from starlette import status

#THIS IS NOT WORKING
@router.get('/categories', status_code=status.HTTP_200_OK)
async def category(category: str, db: Session = Depends(get_db)):
    return await get_products_by_category(category, db)
```

```py
#/src/api/routes/categories.py
from fastapi import APIRouter, Depends
from starlette import status
from sqlalchemy.orm import Session

from src.database.db import get_db
from src.dependencies.auth_deps import verify_google_token_dependency
from src.lib.helpers import get_products_by_category

router = APIRouter(dependencies=[Depends(verify_google_token_dependency)])

#THIS IS WORKING
@router.get('/', status_code=status.HTTP_200_OK)
async def category(category: str, db: Session = Depends(get_db)):
    return await get_products_by_category(category, db)
```

```py
#/src/lib/helpers.py

from src.database.models.products import ProductModel

async def get_products_by_category(category: str, db: Session):
    return db.query(ProductModel).filter(ProductModel.main_category == category).order_by(
        ProductModel.discount.desc()).all()
```

### Description

* Open Postman / swagger and try to execute `http://localhost:8000/api/v1/products/categories?category=Кухня` fails with
```
{
    "detail": "Invalid item ID format"
}
```

*Open Postman / swagger and execute `http://localhost:8000/api/v1/categories?category=Кухня` works as intended

### Operating System

Linux

### Operating System Details

Fedora 40 KDE Plasma

### FastAPI Version

0.111.0

### Pydantic Version

2.7.4

### Python Version

Python 3.11.6

### Additional Context

I've tried to decode the query param since it is comming in cyrilic but the result is still the same. Tried with try-except but it wont even go to the try statement.

**Comment by gustavosett** [comment]
1. Is there any APIRouter at `src/api/routes/products.py`? Check for any typos or misconfigurations in the path.
2. Ensure if UTF-8 encoding is properly configured in both your server and client, as the parameter uses cyrillic characters.
3. Debug the application by printing out the incoming query parameter directly in the route function to see what value is being received.

**Reply by chavdar94** [reply]
Yes there is a router and there sre two more routes in that file that are working correctly and if i pass the category quey param to one of the other two it is finding it and working correctly. Currently im using the route from /api/categories without any other setup snd it is also working. Also i cant debug it since it is not even getting indide the route so i can check whats happening.

**Reply by gustavosett** [reply]
The issue might lie in the Depends(get_db) dependency if it’s failing before reaching the route. Simplify the route by temporarily removing the dependency:
```py
@router.get('/categories', status_code=status.HTTP_200_OK)
async def category(category: str):
    return {"category": category}
```

**Reply by chavdar94** [reply]
It not the db. This is one of the other routes in the products.py file:
```py
@router.get("/", status_code=status.HTTP_200_OK)
async def get_products(
        page: int = Query(1, ge=1),
        page_size: int = Depends(validate_page_size),
        sort: SortOptionEnum = Query(SortOptionEnum.on_sale),
        search: str = Query(None),
        db: Session = Depends(get_db),
) -> ProductResponse:
    """
    Retrieves a list of stores from the cached database.

    Args:
        page (int, optional): The page number. Defaults to 1.
        page_size (int, optional): The number of items per page. Defaults to 10 max 100.
        sort (SortOptionEnum, optional): The sort option. Defaults to SortOptionEnum.relevance.
        db (Session, optional): The database session. Defaults to Depends(get_db).
        search (str, optional): The search string to search for. Defaults to None.:

    Returns:
        StoreResponse: A response containing the stores and pagination information.
    """

    products, total_count, categories = CacheStores.get_cached_stores(
        db, page, page_size, sort, search
    )
    total_pages = (total_count + page_size - 1) // page_size

    pagination = Pagination(
        currentPage=page,
        pageSize=page_size,
        totalPages=total_pages,
        totalResults=total_count,
    )

    sorts = [
        {
            "code": sort_option.value,
            "name": sort_option.display_name,
            "selected": sort == sort_option,
        }
        for sort_option in SortOptionEnum
    ]
    return ProductResponse(
        pagination=pagination, sorts=sorts, products=products, categories=categories
    )
 ```
    
If i pass the same qury param `category: str` to it it is finding it. THe problem is in that exact route.

**Comment by chavdar94** [comment]
I found the problem... Since i have a dynamic that is before the `/caetgories` route it is matching the dynamic route and the error is comming from the validators in the dynamic route.

![image](https://github.com/user-attachments/assets/7fbf28c0-08fc-44f8-9a49-294ebabfae50)
If i move the `/categories` route above the `{/product_id}` route it works fine.
